### ISSUE TITLE
[ATTACK] P2P pubsub spam

### ISSUE THREAD

**Issue by SebastianJ** [ElrondNetwork/elrond-go#1973]
### TL;DR: ###

- Ran my p2p spamming tool on the post-purge network running on version v1.0.132
- Meta stalled for 15 minutes at most, shard 0 / 1 stalled for up to about 6-7 minutes - no shards got permanently stuck.
- About 30-35 validators went offline during the attack
- Multiple reports about node ratings dropping

### Tool: ###
https://github.com/SebastianJ/elrond-peer-attacker

#### Installation: ####
```
mkdir -p $GOPATH/src/github.com/SebastianJ
cd $GOPATH/src/github.com/SebastianJ
git clone https://github.com/SebastianJ/elrond-sdk
git clone https://github.com/SebastianJ/elrond-peer-attacker.git
cd elrond-peer-attacker
./scripts/install.sh
```

#### Usage: ####

Add wallet.pem files to keys/

**Regular tx spamming:**
```
cd data && mv tx_data.txt data.txt && mv tx_receivers.txt receivers.txt && cd ..
./dist/spam --receivers data/receivers.txt --config config/p2p.toml --economics-config config/economics.toml --data data/data.txt --wallets keys --method transactions --concurrency 10
```

**Smart contract spamming:**
```
cd data && mv sc_data.txt data.txt && mv sc_receivers.txt receivers.txt && cd ..
./dist/spam --receivers data/receivers.txt --config config/p2p.toml --economics-config config/economics.toml --data data/data.txt --wallets keys --method transactions --concurrency 10
```

**Heartbeat spamming:**
```
./dist/spam --receivers data/receivers.txt --config config/p2p.toml --economics-config config/economics.toml --data data/data.txt --wallets keys --method heartbeats --concurrency 100
```

_(And yeah, the code is quite a mess, I merged the current p2p spam functionality with a previous repo + did a bunch of experiments all throughout coding everything)_

### Effectiveness: ###
A lot less compared to what I anticipated.

Just a couple of hours earlier (on v1.0.131) shards 0 / 1 got permanently stuck while I ran p2p tx spam. Before that metachain had already gotten messed up (presumably from a previous set of meta shard spam tests I did - but not 100% verifiable).

**The attack is quite cheap to carry out, it roughly costs me €2/day to run it** using two VPS servers and I can obviously scale it up using more servers so that it propagates more traffic.

I can also just leave it running, I wouldn't be surprised if there eventually could be long-lasting effects (validators getting jailed, shards getting stuck etc) if just letting the tool run for a prolonged period of time.

This attack vector also utilizes other nodes to help out with spamming since the actual messages are propagated via pubsub, something which also makes it harder for noder runners to identify the true source/originator of the spam.

### Screenshots ###
Offline nodes due to attack:
![Screenshot 2020-06-17 at 23 21 11](https://user-images.githubusercontent.com/13743/84956429-77b64280-b0f9-11ea-947d-d96d0fb91ffe.png)

Example of node ratings dropping:
![2020-06-17 22 31 03](https://user-images.githubusercontent.com/13743/84956735-1b9fee00-b0fa-11ea-9ccd-5c8047e4986d.jpg)

Running attack earlier on v1.0.131 (i.e. much more effective - good job on v1.0.132!):
![Screenshot 2020-06-17 at 11 14 39](https://user-images.githubusercontent.com/13743/84956472-93214d80-b0f9-11ea-96ab-3cb7d2a58d82.png)
![Screenshot 2020-06-17 at 11 14 45](https://user-images.githubusercontent.com/13743/84956484-97e60180-b0f9-11ea-9a25-0cbb61a5df8c.png)
